base_layer/core/src/base_node/states/horizon_state_sync/state_sync.rs

```diff
@@ -0,0 +1,705 @@
+        BaseNodeStateMachine,
+    },
+    blocks::BlockHeader,
+    chain_storage::{async_db, BlockchainBackend, BlockchainDatabase, ChainMetadata, MmrTree},
+    transactions::{
+        transaction::{TransactionKernel, TransactionOutput},
+        types::HashOutput,
+    },
+};
+use croaring::Bitmap;
+use log::*;
+use std::cmp;
+use tari_comms::peer_manager::NodeId;
+use tari_crypto::tari_utilities::Hashable;
+use tokio::task::spawn_blocking;
+
+const LOG_TARGET: &str = "c::bn::states::horizon_state_sync";
+
+#[derive(Clone, Debug)]
+pub struct HorizonStateSync {
+    network_metadata: ChainMetadata,
+    sync_peers: Vec<NodeId>,
+}
+
+impl HorizonStateSync {
+    pub fn new(network_metadata: ChainMetadata, sync_peers: Vec<NodeId>) -> Self {
+        Self {
+            network_metadata,
+            sync_peers,
+        }
+    }
+
+    pub async fn next_event<B: BlockchainBackend + 'static>(
+        &mut self,
+        shared: &mut BaseNodeStateMachine<B>,
+    ) -> StateEvent
+    {
+        match async_db::get_metadata(shared.db.clone()).await {
+            Ok(local_metadata) => {
+                shared
+                    .set_status_info(StatusInfo::HorizonSync(BlockSyncInfo::new(
+                        Some(self.network_metadata.height_of_longest_chain()),
+                        Some(local_metadata.height_of_longest_chain()),
+                        Some(self.sync_peers.clone()),
+                    )))
+                    .await;
+
+                if !local_metadata.is_pruned_node() {
+                    warn!(
+                        target: LOG_TARGET,
+                        "HorizonStateSync invoked but node is not in pruned mode"
+                    );
+                    return StateEvent::HorizonStateSynchronized;
+                }
+
+                info!(target: LOG_TARGET, "Synchronizing horizon state.");
+                let horizon_sync_height = self.get_horizon_sync_height(&shared, &local_metadata);
+                let local_tip_height = local_metadata.height_of_longest_chain.unwrap_or(0);
+                if local_tip_height >= horizon_sync_height {
+                    debug!(target: LOG_TARGET, "Horizon state already synchronized.");
+                    return StateEvent::HorizonStateSynchronized;
+                }
+                debug!(
+                    target: LOG_TARGET,
+                    "Horizon sync starting to height {}", horizon_sync_height
+                );
+
+                let mut horizon_header_sync = HorizonStateSynchronization {
+                    shared,
+                    local_metadata,
+                    sync_peers: &mut self.sync_peers,
+                    horizon_sync_height,
+                };
+                match horizon_header_sync.synchronize().await {
+                    Ok(()) => {
+                        info!(target: LOG_TARGET, "Horizon state has synchronised.");
+                        StateEvent::HorizonStateSynchronized
```

Comment:
I might have missed it, but maybe we need to add a loop around this to handle multiple horizons, or perhaps return a different status to return to this state for the next horizon before moving to BlockSync

Reply:
Not sure I follow - but I believe that the `horizon_sync_height_offset` is meant to help with this problem - might need to discuss the strategy some more